rust/worker/src/execution/operators/merge_metadata_results.rs

```diff
@@ -107,289 +113,124 @@ impl Operator<MergeMetadataResultsOperatorInput, MergeMetadataResultsOperatorOut
-                                );
-                            }
-                        }
-                    }
-                };
+        // Hydrate the remaining data from the record segment
+        if let Some(reader) = record_segment_reader {
+            for merged_id in merged_ids.difference(&logged_offset_ids) {
+                let user_id = reader
+                    .get_user_id_for_offset_id(*merged_id)
+                    .await
+                    .map_err(|e| {
+                        error!("Error reading record segment: {}", e);
+                        MergeMetadataResultsOperatorError::RecordSegmentReadError
+                    })?;
+                ids.push(user_id.to_string());
                 if input.include_metadata {
-                    let data = match record_segment_reader.get_all_data().await {
-                        Ok(data) => data,
-                        Err(e) => {
-                            tracing::error!(
-                                "[Mergemetadata]: Error reading Record Segment: {:?}",
-                                e
-                            );
-                            return Err(MergeMetadataResultsOperatorError::RecordSegmentReadError);
-                        }
-                    };
-                    for record in data.iter() {
-                        // Ignore records processed from the log.
-                        if ids_in_log.contains(record.id) {
-                            continue;
-                        }
-                        ids.push(record.id.to_string());
-                        metadata.push(record.metadata.clone());
-                        match record.document {
-                            Some(document) => documents.push(Some(document.to_string())),
-                            None => documents.push(None),
-                        }
-                    }
-                } else {
-                    let user_ids = match record_segment_reader.get_all_user_ids().await {
-                        Ok(ids) => ids,
-                        Err(e) => {
-                            tracing::error!(
-                                "[Mergemetadata]: Error reading Record Segment: {:?}",
-                                e
-                            );
-                            return Err(MergeMetadataResultsOperatorError::RecordSegmentReadError);
-                        }
-                    };
-                    for id in user_ids {
-                        // Ignore records processed from the log.
-                        if ids_in_log.contains(id) {
-                            continue;
-                        }
-                        ids.push(id.to_string());
-                    }
+                    let record = reader
+                        .get_data_for_offset_id(*merged_id)
+                        .await
+                        .map_err(|e| {
+                            error!("Error reading Record Segment: {}", e);
```

Comment:
nit: prefer tracing::error! as its what we do in all our code.

Reply:
Fixed.